--- attrafax-0.9/compile.sh.orig	2010-08-24 13:36:58.890175592 +0930
+++ attrafax-0.9/compile.sh	2010-08-24 13:37:27.200485604 +0930
@@ -25,18 +25,21 @@
 	elif [ `uname -s` = "SunOS" ]; then
 		CC="/opt/sunstudio12.1/bin/cc"
 	fi
+else
+        CC="$CC"
 fi
 
 if [ -z "$CFLAGS" ]; then
-	CFLAGS="-g"
+	CFLAGS="-std=gnu99  -g -ffast-math -mfast-fp"
 fi
 
 if [ `uname -s` = "SunOS" ]; then
 	CFLAGS="${CFLAGS} -KPIC -D_FILE_OFFSET_BITS=64 -xc99=all"
 else
-	CFLAGS="${CFLAGS} -Wall -Werror -fPIC -D_FILE_OFFSET_BITS=64"
+#	CFLAGS="${CFLAGS} -Wall -Werror -fPIC -D_FILE_OFFSET_BITS=64"
+	CFLAGS="${CFLAGS} -Wall -fPIC -D_FILE_OFFSET_BITS=64"
 fi
-INCLUDE="-I../../include -I/usr/local/include $EXTINCLUDES"
+INCLUDE="-I../../include -I$1 $EXTINCLUDES"
 BLIBS="-lm -lz -lssl -lcrypto"
 
 
@@ -69,11 +72,11 @@
 	fi
 fi
 
-LIBS="-L../../bin -L/usr/local/lib -lt30"
+LIBS="-L../../bin -L$2 -lt30"
 
 BIN_COMMON="rev.o hdlc.o llist.o tlogger.o context.o"
 TONE_COMMON="dds.o goertzel.o cng.o ced.o"
-MODEM_COMMON="eq.o meter.o"
+MODEM_COMMON="eq.o meter.o fastmath.o"
 V21="v21tx.o v21rx.o"
 V27TER="v27tertx2400.o v27terrx2400.o v27tertx4800.o v27terrx4800.o"
 V29="v29tx.o v29rx.o"
@@ -109,11 +112,11 @@
 	${CC} ${CFLAGS} ${INCLUDE} -c "$i"
 done
 
-${CC} ${CFLAGS} -shared -Xlinker ${BLIBS} \
+${CC} ${CFLAGS} -shared -Xlinker ${BLIBS} ${LIBS} \
 	${BIN_COMMON} ${TONE_COMMON} ${MODEMS} ${FAX_COMMON} \
-	${TIFF} ${JPEG} -o ../../${BINDIR}/libt30.so
+	${TIFF} -o ../../${BINDIR}/libt30.so
 
 
-ar rv ../../${BINDIR}/libt30.a ${BIN_COMMON} ${TONE_COMMON} ${MODEMS} ${FAX_COMMON}
-ranlib ../../${BINDIR}/libt30.a
+bfin-linux-uclibc-ar rv ../../${BINDIR}/libt30.a ${BIN_COMMON} ${TONE_COMMON} ${MODEMS} ${FAX_COMMON}
+bfin-linux-uclibc-ranlib ../../${BINDIR}/libt30.a
 
