--- uClinux-dist/linux-2.6.x/arch/blackfin/mach-bf537/boards/Makefile	2009-02-04 16:23:50.000000000 +0200
+++ uClinux-dist/linux-2.6.x/arch/blackfin/mach-bf537/boards/Makefile_	2009-07-14 12:45:03.000000000 +0300
@@ -3,6 +3,7 @@
 #
 
 obj-$(CONFIG_BFIN537_STAMP)            += stamp.o
+obj-$(CONFIG_PR1_APPLIANCE)            += pr1_appliance.o
 obj-$(CONFIG_BFIN537_BLUETECHNIX_CM)   += cm_bf537.o
 obj-$(CONFIG_BFIN537_BLUETECHNIX_TCM)  += tcm_bf537.o
 obj-$(CONFIG_PNAV10)                   += pnav10.o
