--- dahdi_orig	2011-10-21 14:36:33.000000000 -0200
+++ dahdi	2011-10-21 14:38:28.000000000 -0200
@@ -1,8 +1,10 @@
 #!/bin/hush
 # Start up file for DAHDI
+
+SIM_pin=`grep simpin /persistent/etc/asterisk/guipreferences.conf | sed 's/simpin = */\1/'`
 	
 case $1 in
-        start)   modprobe wcfxs opermode=`sed -n 's/^opermode *= *//p' /etc/asterisk/guipreferences.conf`;
+        start)   modprobe wcfxs opermode=`sed -n 's/^opermode *= *//p' /etc/asterisk/guipreferences.conf` SIM_pin=$SIM_pin;
                  if dahdi_scan | grep -q "Blackfin IP0x Board"; then; else; modprobe dahdi_dummy; fi;
                  dahdi_cfg;;
         stop)    modprobe -r wcfxs dahdi_dummy;;
