--- linux/drivers/dahdi/Kconfig	2009-11-29 22:44:33.000000000 +0200
+++ linux/drivers/dahdi/Kconfig	2009-11-29 22:45:25.000000000 +0200
@@ -72,7 +72,7 @@
 config DAHDI_WCTDM
 	tristate "Digium Wildcard TDM400P Support"
 	depends on DAHDI && PCI
-	default DAHDI
+	default "n"
 	---help---
 	  This driver provides support for the Digium Wildcard TDM400P.
 
--- linux/Makefile	2010-03-05 15:47:32.000000000 +0200
+++ linux/Makefile	2010-03-05 15:47:59.000000000 +0200
@@ -88,7 +88,7 @@
 	fi
 	@rm -f $@.tmp
 
-prereq: include/dahdi/version.h firmware-loaders
+prereq: include/dahdi/version.h
 
 stackcheck: checkstack modules
 	./checkstack kernel/*.ko kernel/*/*.ko
--- tools/dahdi_scan.c	2009-11-04 18:36:21.000000000 +0200
+++ tools/dahdi_scan.c	2010-05-23 12:36:32.000000000 +0300
@@ -156,8 +156,11 @@
 			fprintf(stdout, "framing=");
 			if (s.lineconfig & DAHDI_CONFIG_ESF) fprintf(stdout, "ESF");
 			else if (s.lineconfig & DAHDI_CONFIG_D4) fprintf(stdout, "D4");
-			else if (s.lineconfig & DAHDI_CONFIG_CCS) fprintf(stdout, "CCS");
-			else if (s.lineconfig & DAHDI_CONFIG_CRC4) fprintf(stdout, "/CRC4");
+			else
+                        if (s.lineconfig & DAHDI_CONFIG_CCS) {
+                                fprintf(stdout, "CCS");
+                                if (s.lineconfig & DAHDI_CONFIG_CRC4) fprintf(stdout, "/CRC4"); //Penev in general CRC4 can go together with CAS as well
+                        }	
 			fprintf(stdout, "\n");
 		} else {
 			/* this is an analog span */
--- linux/drivers/dahdi/Kbuild	2010-06-11 11:06:41.000000000 +0300
+++ linux/drivers/dahdi/Kbuild	2010-06-11 11:06:56.000000000 +0300
@@ -1,5 +1,5 @@
 obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI)			+= dahdi.o
-#obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI_DUMMY)		+= dahdi_dummy.o
+obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI_DUMMY)		+= dahdi_dummy.o
 obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI_DYNAMIC)		+= dahdi_dynamic.o
 obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI_DYNAMIC_LOC)	+= dahdi_dynamic_loc.o
 obj-$(DAHDI_BUILD_ALL)$(CONFIG_DAHDI_DYNAMIC_ETH)	+= dahdi_dynamic_eth.o
@@ -33,12 +33,14 @@
 
 # Only enable this if you think you know what you're doing. This is not
 # supported yet:
-#obj-m += dahdi_echocan_oslec.o
+obj-m += dahdi_echocan_oslec.o
 #
 # A quick and dirty way to build OSLEC, if you happened to place it
 # yourself in the dahdi source tree. This is experimental. See README
 # regarding OSLEC.
-#obj-m += ../staging/echo/
+obj-m += ../staging/echo/
+
+obj-m += dahdi_echocan_zarlink.o
 
 CFLAGS_MODULE += -I$(DAHDI_INCLUDE) -I$(src)
 
