--- uClinux-dist/linux-2.6.x/drivers/mtd/nand/Kconfig	2009-07-14 12:39:00.000000000 +0300
+++ uClinux-dist/linux-2.6.x/drivers/mtd/nand/Kconfig_	2009-07-14 12:55:45.000000000 +0300
@@ -37,10 +37,10 @@
 	  of these chips were reused by later, larger chips.
 
 config MTD_NAND_BFIN
-	tristate "NAND Flash device for BF537 STAMP board (DEPRECATED)"
+	tristate "NAND Flash device for Blackfin board"
 	depends on BLACKFIN && MTD_NAND && MTD_PARTITIONS
 	help
-	  This enables the driver for the NAND flash ST-NAND123W3A 
+	  This enables the driver for the NAND flash 
 	  for Blackfin Processors
 
 	  This is deprecated.  Please use the generic platform NAND
@@ -49,10 +49,20 @@
 config BFIN_NAND_BASE
 	hex "NAND Flash Base Address"
 	depends on MTD_NAND_BFIN
-	default 0x20212000
+	default 0x20000000
 	help
 	  NAND Flash Base Address
 
+config BFIN_NAND_SIZE
+       hex "NAND Flash Size"
+       depends on MTD_NAND_BFIN
+       default 0x10000000
+       help
+         NAND Flash size, for example 0x10000000 for 256Mbyte, or
+          0x02000000 for 32MByte.  The NAND will be divided into
+         two partitions.  The first partition is 4M (kernel), the
+         second partition the remaining space (e.g. yaffs).
+
 config BFIN_NAND_CLE
 	int "NAND Flash Command Latch Enable (CLE) Address strobe A[x]"
 	depends on MTD_NAND_BFIN
--- uClinux-dist/linux-2.6.x/drivers/mtd/nand/bfin_nand.c	2007-02-06 12:41:48.000000000 -0500
+++ uClinux-dist/linux-2.6.x/drivers/mtd/nand/bfin_nand.c	2007-10-26 10:39:01.000000000 -0400
@@ -54,21 +54,26 @@
  */
 const static struct mtd_partition partition_info[] = {
 	{
-		.name = "linux kernel",
+		.name = "uImage",
 		.offset = 0,
-		.size = 4 *1024*1024,
+		.size = 10 *1024*1024,
 	},
+	{
+                .name = "uImage backup",
+                .offset = 10 * 1024 * 1024,
+                .size = 10 *1024*1024,
+        },
 #ifdef CONFIG_PNAV10 /* 1G x 8 NAND Flash */
 	{
 		.name = "file system",
-		.offset = 4 *1024*1024,
+		.offset = 10 *1024*1024,
 		.size = (1024-4) *1024*1024,
 	}
 #else
 	{
-		.name = "file system",
-		.offset = 4 *1024*1024,
-		.size = (128-4) *1024*1024,
+		.name = "persistent file system",
+		.offset = 20 *1024*1024,
+		.size = (CONFIG_BFIN_NAND_SIZE-(20 *1024*1024))
 	}
 #endif
 };
