--- src/config.h	2010-02-14 21:30:56.000000000 +0200
+++ src/config.h	2010-02-14 21:31:12.000000000 +0200
@@ -304,10 +304,10 @@
 #endif
 
 /* Define to rpl_malloc if the replacement function should be used. */
-#define malloc rpl_malloc
+//#define malloc rpl_malloc
 
 /* Define to rpl_realloc if the replacement function should be used. */
-#define realloc rpl_realloc
+//#define realloc rpl_realloc
 
 /* Define to empty if the keyword `volatile' does not work. Warning: valid
    code using `volatile' can become incorrect without. Disable with care. */
