--- blackfin-linux-dist/user/blkfin-apps/Makefile.orig	2010-07-26 11:58:59.000000000 +0100
+++ blackfin-linux-dist/user/blkfin-apps/Makefile	2011-06-10 12:49:19.706438387 +0100
@@ -24,7 +24,10 @@
 dir_$(CONFIG_USER_EDN_BENCH)                     += edn
 dir_$(CONFIG_USER_DSP_BRIDGE)                    += dsp-bridge
 
-all %::
+all::
+	for i in $(dir_y) ; do $(MAKE) -C $$i $@ || exit $$? ; done
+
+%::
 	for i in $(dir_y) ; do $(MAKE) -C $$i $@ || exit $$? ; done
 
 clean:
--- blackfin-linux-dist/user/busybox/busybox-1.16.2/Makefile.orig	2011-06-10 12:52:02.208111912 +0100
+++ blackfin-linux-dist/user/busybox/busybox-1.16.2/Makefile	2011-06-10 12:58:38.906433389 +0100
@@ -428,7 +428,12 @@
 -include $(srctree)/arch/$(ARCH)/Makefile
 export KBUILD_DEFCONFIG
 
-config %config: scripts_basic outputmakefile FORCE
+config: scripts_basic outputmakefile FORCE
+	$(Q)mkdir -p include
+	$(Q)$(MAKE) $(build)=scripts/kconfig $@
+	$(Q)$(MAKE) -C $(srctree) KBUILD_SRC= .kernelrelease
+
+%config: scripts_basic outputmakefile FORCE
 	$(Q)mkdir -p include
 	$(Q)$(MAKE) $(build)=scripts/kconfig $@
 	$(Q)$(MAKE) -C $(srctree) KBUILD_SRC= .kernelrelease
@@ -1276,7 +1281,10 @@
 	$(Q)$(MAKE) $(build)=$(build-dir) $(target-dir)$(notdir $@)
 
 # Modules
-/ %/: prepare scripts FORCE
+/: prepare scripts FORCE
+	$(Q)$(MAKE) KBUILD_MODULES=$(if $(CONFIG_MODULES),1) \
+	$(build)=$(build-dir)
+%/: prepare scripts FORCE
 	$(Q)$(MAKE) KBUILD_MODULES=$(if $(CONFIG_MODULES),1) \
 	$(build)=$(build-dir)
 %.ko: prepare scripts FORCE
--- blackfin-linux-dist/vendors/Makefile.orig	2011-06-10 13:05:40.612398073 +0100
+++ blackfin-linux-dist/vendors/Makefile	2011-06-10 13:05:19.060706501 +0100
@@ -25,7 +25,10 @@
 
 # Note: the all and clean targets are not normally used,
 # since user/Makefile jumps directory to the product directory
-all image clean romfs romfs.post vendor_%:
+all image clean romfs romfs.post:
+	$(MAKE) -C $(dir_v) dir_v=$(dir_v) -f $(VEND)/vendors-common.mak $@
+
+vendor_%:
 	$(MAKE) -C $(dir_v) dir_v=$(dir_v) -f $(VEND)/vendors-common.mak $@
 
 prune:
--- blackfin-linux-dist/user/busybox/busybox-1.16.2/scripts/mkmakefile.orig	2011-06-10 13:10:58.489021840 +0100
+++ blackfin-linux-dist/user/busybox/busybox-1.16.2/scripts/mkmakefile	2011-06-10 13:10:38.752289198 +0100
@@ -31,6 +31,9 @@
 
 Makefile:;
 
-\$(filter-out all Makefile,\$(MAKECMDGOALS)) %/:
+\$(filter-out all Makefile,\$(MAKECMDGOALS)):
+	\$(MAKE) -C \$(KERNELSRC) O=\$(KERNELOUTPUT) \$@
+
+%/:
 	\$(MAKE) -C \$(KERNELSRC) O=\$(KERNELOUTPUT) \$@
 EOF
