--- uClinux-dist.R1.1-RC3/linux-2.6.x/drivers/mtd/nand/bfin_nand.c	2007-02-06 12:41:48.000000000 -0500
+++ uClinux-dist.R1.1-RC3/linux-2.6.x/drivers/mtd/nand/bfin_nand.c	2007-10-26 10:39:01.000000000 -0400
@@ -56,19 +56,19 @@
 	{
 		.name = "linux kernel",
 		.offset = 0,
-		.size = 4 *1024*1024,
+		.size = 8 *1024*1024,
 	},
 #ifdef CONFIG_PNAV10 /* 1G x 8 NAND Flash */
 	{
 		.name = "file system",
-		.offset = 4 *1024*1024,
+		.offset = 8 *1024*1024,
 		.size = (1024-4) *1024*1024,
 	}
 #else
 	{
 		.name = "file system",
-		.offset = 4 *1024*1024,
-		.size = (128-4) *1024*1024,
+		.offset = 8 *1024*1024,
+		.size = (CONFIG_BFIN_NAND_SIZE-(8 *1024*1024))
 	}
 #endif
 };
--- uClinux-dist.R1.1-RC3/linux-2.6.x/drivers/mtd/nand/Kconfig	2007-10-26 12:34:25.000000000 -0400
+++ uClinux-dist.R1.1-RC3/linux-2.6.x/drivers/mtd/nand/Kconfig	2007-10-26 12:34:59.000000000 -0400
@@ -2,7 +2,7 @@
 # $Id: Kconfig 2632 2007-01-11 12:13:09Z cooloney $
 
 menu "NAND Flash Device Drivers"
-	depends on MTD!=n
+	depends on MTD != n
 
 config MTD_NAND
 	tristate "NAND Device Support"
@@ -32,19 +32,29 @@
 	  The original Linux implementation had byte 0 and 1 swapped.
 
 config MTD_NAND_BFIN
-	tristate "NAND Flash device for BF537 STAMP board"
+	tristate "NAND Flash device for Blackfin"
 	depends on BFIN && MTD_NAND && MTD_PARTITIONS
 	help
-	  This enables the driver for the NAND flash ST-NAND123W3A 
-	  for Blackfin Processors
+	  This enables the driver for the NAND flash for Blackfin 
+          Processors.
 
 config BFIN_NAND_BASE
 	hex "NAND Flash Base Address"
 	depends on MTD_NAND_BFIN
-	default 0x20212000
+	default 0x20000000
 	help
 	  NAND Flash Base Address
 
+config BFIN_NAND_SIZE
+	hex "NAND Flash Size"
+	depends on MTD_NAND_BFIN
+	default 0x10000000
+	help
+	  NAND Flash size, for example 0x10000000 for 256Mbyte, or
+          0x02000000 for 32MByte.  The NAND will be divided into
+	  two partitions.  The first partition is 4M (kernel), the 
+	  second partition the remaining space (e.g. yaffs).
+
 config BFIN_NAND_CLE
 	int "NAND Flash Command Latch Enable (CLE) Address strobe A[x]"
 	depends on MTD_NAND_BFIN
