diff -ur asterisk/channels/chan_dahdi.c asterisk-gsm1/channels/chan_dahdi.c
--- asterisk/channels/chan_dahdi.c	2010-09-06 11:36:34.000000000 +0300
+++ asterisk-gsm1/channels/chan_dahdi.c	2010-09-06 11:34:09.000000000 +0300
@@ -7047,7 +7047,27 @@
 				dahdi_setlinear(p->subs[index].dfd, p->subs[index].linear);
 #if 1
 				restore_gains(p);
-#endif				
+#endif			
+				#define WCGSM_GET_CALLERID    10
+				if (!number) { //Get Caller ID from the GSM1 module
+					       //At this point we are not sure if this is GSM1 channel but the worst case we 
+					       //will find that the ioctl(/dev/dahdi/wcgsm,callerid) will return empty 
+					int wcgsm_fd;
+					char gsm1_number[60];
+					int wcgsm_res;
+
+					wcgsm_fd=open("/dev/dahdi/wcgsm", O_RDWR);
+					if(wcgsm_fd){
+						wcgsm_res=ioctl(wcgsm_fd, WCGSM_GET_CALLERID, (long)gsm1_number);
+						if(!wcgsm_res) { 
+							number=gsm1_number;
+							
+							if (option_debug) ast_log(LOG_DEBUG, "GSM1 CallerID number: %s\n", number);
+                                               } 
+						close(wcgsm_fd);					
+					}							 
+				}	
+
 				if (res < 0) {
 					ast_log(LOG_WARNING, "CallerID returned with error on channel '%s'\n", chan->name);
 				}
