--- Makefile	2011-02-01 02:37:49.000000000 +0200
+++ Makefile	2011-08-05 21:46:02.000000000 +0300
@@ -24,11 +24,19 @@
 # Uncomment if you want libpri not send PROGRESS_INDICATOR w/ALERTING
 #ALERTING=-DALERTING_NO_PROGRESS
 
-CC=gcc
+CC ?= bfin-linux-uclibc-gcc
+CCFLAGS = $(CFLAGS) -c
+INSTALL=install
+AR ?= bfin-linux-uclibc-ar
+RANLIB ?= bfin-linux-uclibc-ranlib
+STRIP ?= bfin-linux-uclibc-strip
+LDCONFIG_FLAGS=
+
+
 GREP=grep
 AWK=awk
 
-OSARCH=$(shell uname -s)
+OSARCH=$(ARCH)
 PROC?=$(shell uname -m)
 
 # SONAME version; should be changed on every ABI change
@@ -90,7 +98,7 @@
 	rose_qsig_mwi.lo \
 	rose_qsig_name.lo \
 	version.lo
-CFLAGS=-Wall -Werror -Wstrict-prototypes -Wmissing-prototypes -g -fPIC $(ALERTING) $(LIBPRI_OPT) $(COVERAGE_CFLAGS)
+CFLAGS= -mfdpic -fomit-frame-pointer -Wall -Werror -Wstrict-prototypes -Wmissing-prototypes -g $(ALERTING) $(COVERAGE_CFLAGS) $(LIBPRI_OPT)
 INSTALL_PREFIX=$(DESTDIR)
 INSTALL_BASE=/usr
 libdir?=$(INSTALL_BASE)/lib
@@ -101,7 +109,7 @@
   endif
   LDCONFIG=/usr/bin/true
 else
-  SOFLAGS=-shared -Wl,-h$(DYNAMIC_LIBRARY) $(COVERAGE_LDFLAGS)
+  SOFLAGS=-mfdpic -shared -Wl,-h$(DYNAMIC_LIBRARY) $(COVERAGE_LDFLAGS)
   LDCONFIG = /sbin/ldconfig
 endif
 ifneq (,$(findstring X$(OSARCH)X, XLinuxX XGNU/kFreeBSDX XGNUX))
@@ -145,7 +153,7 @@
 SOFLAGS += -m32
 endif
 
-all: $(STATIC_LIBRARY) $(DYNAMIC_LIBRARY)
+all: $(STATIC_LIBRARY) $(DYNAMIC_LIBRARY) ar-target
 
 update:
 	@if [ -d .svn ]; then \
@@ -164,23 +172,15 @@
 		echo "Not under version control";  \
 	fi
 
-install: $(STATIC_LIBRARY) $(DYNAMIC_LIBRARY)
-	mkdir -p $(INSTALL_PREFIX)$(libdir)
-	mkdir -p $(INSTALL_PREFIX)$(INSTALL_BASE)/include
-ifneq (${OSARCH},SunOS)
-	install -m 644 libpri.h $(INSTALL_PREFIX)$(INSTALL_BASE)/include
-	install -m 755 $(DYNAMIC_LIBRARY) $(INSTALL_PREFIX)$(libdir)
-	#if [ -x /usr/sbin/sestatus ] && ( /usr/sbin/sestatus | grep "SELinux status:" | grep -q "enabled"); then /sbin/restorecon -v $(INSTALL_PREFIX)$(libdir)/$(DYNAMIC_LIBRARY); fi
-	( cd $(INSTALL_PREFIX)$(libdir) ; ln -sf libpri.so.$(SONAME) libpri.so)
-	install -m 644 $(STATIC_LIBRARY) $(INSTALL_PREFIX)$(libdir)
-	if test $$(id -u) = 0; then $(LDCONFIG) $(LDCONFIG_FLAGS) $(INSTALL_PREFIX)$(libdir); fi
-else
-	install -f $(INSTALL_PREFIX)$(INSTALL_BASE)/include -m 644 libpri.h
-	install -f $(INSTALL_PREFIX)$(libdir) -m 755 $(DYNAMIC_LIBRARY)
-	( cd $(INSTALL_PREFIX)$(libdir) ; ln -sf libpri.so.$(SONAME) libpri.so)
-	install -f $(INSTALL_PREFIX)$(libdir) -m 644 $(STATIC_LIBRARY)
-endif
-
+romfs: all 
+	[ -d $(ROMFSDIR)/lib ] || mkdir -p $(ROMFSDIR)/lib ; \
+	$(ROMFSINST) $(DYNAMIC_LIBRARY) /lib/$(DYNAMIC_LIBRARY) ; \
+	( cd $(ROMFSDIR)/lib ; ln -sf $(DYNAMIC_LIBRARY) libpri.so ) \
+
+ar-target: 
+	$(INSTALL) $(DYNAMIC_LIBRARY) -m 644  $(STAGEDIR)/usr/lib
+	$(INSTALL) libpri.h -m 644  $(STAGEDIR)/usr/include
+	ln -sf $(STAGEDIR)/usr/lib/$(DYNAMIC_LIBRARY) $(STAGEDIR)/usr/lib/libpri.so
 uninstall:
 	@echo "Removing Libpri"
 	rm -f $(INSTALL_PREFIX)$(libdir)/libpri.so.$(SONAME)
@@ -212,12 +212,13 @@
 	$(CC) $(CFLAGS) $(MAKE_DEPS) -c -o $@ $<
 
 $(STATIC_LIBRARY): $(STATIC_OBJS)
-	ar rcs $(STATIC_LIBRARY) $(STATIC_OBJS)
-	ranlib $(STATIC_LIBRARY)
+	$(AR) rcs $(STATIC_LIBRARY) $(STATIC_OBJS)
+	$(RANLIB) $(STATIC_LIBRARY)
+	$(STRIP) $(STATIC_LIBRARY)
 
 $(DYNAMIC_LIBRARY): $(DYNAMIC_OBJS)
 	$(CC) $(SOFLAGS) -o $@ $(DYNAMIC_OBJS)
-	$(LDCONFIG) $(LDCONFIG_FLAGS) .
+	$(STRIP) $(DYNAMIC_LIBRARY)
 	ln -sf libpri.so.$(SONAME) libpri.so
 
 version.c: FORCE
