diff -ruN u-boot-1.1.5-bf1-orig/drivers/nand/nand_base.c u-boot-1.1.5-bf1/drivers/nand/nand_base.c
--- u-boot-1.1.5-bf1-orig/drivers/nand/nand_base.c	2006-11-05 23:41:22.000000000 +0200
+++ u-boot-1.1.5-bf1/drivers/nand/nand_base.c	2011-01-24 15:40:07.000000000 +0200
@@ -2309,6 +2309,7 @@
 
 	/* Select the device */
 	this->select_chip(mtd, 0);
+	this->cmdfunc (mtd, NAND_CMD_RESET, -1, -1);
 
 	/* Send the command for reading device ID */
 	this->cmdfunc (mtd, NAND_CMD_READID, 0x00, -1);
